v1/remote/image_test.go

```diff
@@ -0,0 +1,292 @@
+	}
+
+	img := remoteImage{
+		ref:    mustNewTag(t, fmt.Sprintf("%s/%s:latest", u.Host, expectedRepo)),
+		client: http.DefaultClient,
+	}
+
+	if _, err := img.RawManifest(); err == nil {
+		t.Error("RawManifest() = nil; wanted error")
+	}
+}
+
+func TestRawConfigFileNotFound(t *testing.T) {
+	img := randomImage(t)
+	expectedRepo := "foo/bar"
+	manifestPath := fmt.Sprintf("/v2/%s/manifests/latest", expectedRepo)
+	configPath := fmt.Sprintf("/v2/%s/blobs/%s", expectedRepo, mustConfigName(t, img))
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case configPath:
+			if r.Method != http.MethodGet {
+				t.Errorf("Method; got %v, want %v", r.Method, http.MethodGet)
+			}
+			w.WriteHeader(http.StatusNotFound)
+		case manifestPath:
+			if r.Method != http.MethodGet {
+				t.Errorf("Method; got %v, want %v", r.Method, http.MethodGet)
+			}
+			w.Write(mustRawManifest(t, img))
+		default:
+			t.Fatalf("Unexpected path: %v", r.URL.Path)
+		}
+	}))
+	defer server.Close()
+	u, err := url.Parse(server.URL)
+	if err != nil {
+		t.Fatalf("url.Parse(%v) = %v", server.URL, err)
+	}
+
+	rmt := remoteImage{
+		ref:    mustNewTag(t, fmt.Sprintf("%s/%s:latest", u.Host, expectedRepo)),
+		client: http.DefaultClient,
+	}
+
+	if _, err := rmt.RawConfigFile(); err == nil {
+		t.Error("RawConfigFile() = nil; wanted error")
+	}
+}
+
+func TestImage(t *testing.T) {
+	img := randomImage(t)
+	expectedRepo := "foo/bar"
+	configPath := fmt.Sprintf("/v2/%s/blobs/%s", expectedRepo, mustConfigName(t, img))
+	manifestPath := fmt.Sprintf("/v2/%s/manifests/latest", expectedRepo)
+	manifestReqCount := 0
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case "/v2/":
+			w.WriteHeader(http.StatusOK)
+		case configPath:
+			if r.Method != http.MethodGet {
+				t.Errorf("Method; got %v, want %v", r.Method, http.MethodGet)
+			}
+			w.Write(mustRawConfigFile(t, img))
+		case manifestPath:
+			manifestReqCount++
+			if r.Method != http.MethodGet {
+				t.Errorf("Method; got %v, want %v", r.Method, http.MethodGet)
+			}
+			w.Write(mustRawManifest(t, img))
+		default:
+			t.Fatalf("Unexpected path: %v", r.URL.Path)
+		}
+	}))
+	defer server.Close()
+	u, err := url.Parse(server.URL)
+	if err != nil {
+		t.Fatalf("url.Parse(%v) = %v", server.URL, err)
+	}
+
+	tag := mustNewTag(t, fmt.Sprintf("%s/%s:latest", u.Host, expectedRepo))
+	rmt, err := Image(tag, authn.Anonymous, http.DefaultTransport)
+	if err != nil {
+		t.Errorf("Image() = %v", err)
+	}
+
+	if want, got := string(mustRawManifest(t, img)), string(mustRawManifest(t, rmt)); want != got {
+		t.Errorf("RawManifest() = %v, want %v", got, want)
+	}
+	if want, got := string(mustRawConfigFile(t, img)), string(mustRawConfigFile(t, rmt)); want != got {
+		t.Errorf("RawConfigFile() = %v, want %v", got, want)
+	}
+	// Make sure caching works.
+	if manifestReqCount != 1 {
```

Comment:
I realized partway through that we are able to avoid even a `HEAD` request for `BlobSize` because we pull this from the manifest.  Can you verify that we make no `/v2/foo/blobs/bar` requests for `BlobSize()` too?

Reply:
Done.